core/platform/bulk_email/mailchimp_bulk_email_services.py

```diff
@@ -0,0 +1,171 @@
+    """
+    client = _get_mailchimp_class()
+    subscriber_hash = _get_subscriber_hash(user_email)
+
+    try:
+        client.lists.members.get(
+            feconf.MAILCHIMP_AUDIENCE_ID, subscriber_hash)
+        client.lists.members.delete_permanent(
+            feconf.MAILCHIMP_AUDIENCE_ID, subscriber_hash)
+    except mailchimpclient.MailChimpError as error:
+        # This is has to be done since the message can only be accessed from
+        # MailChimpError by error.message in Python2, but this is deprecated in
+        # Python3.
+        # In Python3, the message can be accessed directly by KeyError
+        # (https://github.com/VingtCinq/python-mailchimp/pull/65), so as a
+        # workaround for Python2, the 'message' attribute is obtained by
+        # str() and then it is converted to dict.
+        error_message = ast.literal_eval(python_utils.UNICODE(error))
+        # Ignore if the error corresponds to "User does not exist".
+        if error_message['status'] != 404:
+            raise Exception(error_message['detail'])
+
+
+def add_or_update_user_status(user_email, can_receive_email_updates):
+    """Subscribes/unsubscribes an existing user or creates a new user with
+    correct status in the mailchimp DB.
+
+    NOTE: Callers should ensure that the user's corresponding
+    UserEmailPreferencesModel.site_updates field is kept in sync.
+
+    Args:
+        user_email: str. Email ID of the user. Email is used to uniquely
+            identify the user in the mailchimp DB.
+        can_receive_email_updates: bool. Whether they want to be subscribed to
+            list or not.
+    """
+    client = _get_mailchimp_class()
+    subscriber_hash = _get_subscriber_hash(user_email)
+
+    subscribed_mailchimp_data = {
+        'email_address': user_email,
+        'status': 'subscribed'
+    }
+
+    unsubscribed_mailchimp_data = {
+        'email_address': user_email,
+        'status': 'unsubscribed'
+    }
+
+    try:
+        member_details = client.lists.members.get(
+            feconf.MAILCHIMP_AUDIENCE_ID, subscriber_hash)
+
+        # If member is already added to mailchimp list, we cannot permanently
+        # delete a list member, since they cannot be programmatically added
+        # back, so we change their status based on preference.
+        if (
+                can_receive_email_updates and
+                member_details['status'] != 'subscribed'):
+            client.lists.members.update(
+                feconf.MAILCHIMP_AUDIENCE_ID, subscriber_hash,
+                subscribed_mailchimp_data)
+        elif (
+                not can_receive_email_updates and
+                member_details['status'] == 'subscribed'):
+            client.lists.members.update(
+                feconf.MAILCHIMP_AUDIENCE_ID, subscriber_hash,
+                unsubscribed_mailchimp_data)
+
+    except mailchimpclient.MailChimpError as error:
+        # This is has to be done since the message can only be accessed from
```

Comment:
Ditto -- drop "is", add comments for py2/py3.

Reply:
Done